### PR TITLE
Libtask Integration

### PR DESCRIPTION
**Pull request by yebai**
Replacement of https://github.com/TuringLang/Turing.jl/pull/1757. Awaiting for a new Libtask [release](https://github.com/TuringLang/Libtask.jl/pull/107) before merging.

### CODE REVIEW

File: test/Project.toml [hunk 1]
```diff
@@ -12,6 +12,7 @@ DynamicPPL = "366bfd00-2699-11ea-058f-f148b4cae6d8"
 FiniteDifferences = "26cc04aa-876d-5657-8c51-4c34ba976000"
 ForwardDiff = "f6369f11-7733-5829-9624-2563aa707210"
 GalacticOptim = "a75be94c-b780-496d-a8a9-0878b188d577"
+Libtask = "6f1fad26-d15e-5dc8-ae53-837a1d7b8c9f"
```
[the remark this answers is not in the document]
Comment (yebai) [reply test/Project.toml:15]: It's added by CompatHelper https://github.com/TuringLang/Turing.jl/pull/1765

[the remark this answers is not in the document]
Comment (devmotion) [reply test/Project.toml:15]: But CompatHelper only added the compact entry. I am surprised there is a Libtask dependency at all - it shows up as a new addition in this PR but I missed any change in the tests themselves that would explain this dependency.

**Comment by yebai** [comment]
With https://github.com/TuringLang/Libtask.jl/pull/107, only one test is failing now!

File: Project.toml [hunk 2]
```diff
@@ -59,4 +59,4 @@ StatsBase = "0.32, 0.33"
 StatsFuns = "0.8, 0.9"
 Tracker = "0.2.3"
 ZygoteRules = "0.2"
-julia = "1.3, 1.4, 1.5, 1.6"
+julia = "1.6, 1.7"
```
Comment (devmotion) [inline Project.toml:62]: This is the same as
```suggestion
julia = "1.6"
```
i.e., Julia 1.6 and all newer versions are supported. (The current compat entry seems weird, I guess it was supposed to bound it to < 1.7 but is incorrect.)

File: test/Project.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (devmotion) [reply test/Project.toml:15]: It seems the entry here was added in https://github.com/TuringLang/Turing.jl/pull/1757 (without compat bound, hence CompatHelper opened a PR). But also there it seems it was not needed, so probably both the entry here and the compat entry below could be removed?
```suggestion
```

File: test/Project.toml [hunk 3]
```diff
@@ -42,6 +43,7 @@ DynamicPPL = "0.17.2"
 FiniteDifferences = "0.10.8, 0.11, 0.12"
 ForwardDiff = "0.10.12"
 GalacticOptim = "0.4, 1, 2"
+Libtask = "0.6"
```
Comment (devmotion) [inline test/Project.toml:46]:
```suggestion
```

**Comment by yebai** [comment]
A few tests are still failing for Julia 1.6 due to tight error bounds. Basically, the random seed that works for Julia 1.7 turns out not to work well for Julia 1.6.  The results are quite close, but still erroring due to current tight error bounds. Maybe we should relax the error a bit threshold for the sensitive tests, or increase the number of samples (which leads to a longer CI time).

**Comment by devmotion** [comment]
Maybe we could use StableRNGs? There was a major change of the RNG in 1.7, and generally random numbers are not guaranteed to be reproducible with different Julia versions, OS, or architectures.

File: .github/workflows/Documentation.yml [hunk 4]
```diff
@@ -15,7 +15,7 @@ jobs:
       - name: Set up Julia
         uses: julia-actions/setup-julia@v1
         with:
-          version: '1.6'
+          version: '1.7'
```
Comment (devmotion) [inline .github/workflows/Documentation.yml:18]: Maybe
```suggestion
          version: '1'
```
so always the stable version is tested and we don't have to adjust CI for Julia 1.8 etc. manually?

File: .github/workflows/DynamicHMC.yml [hunk 5]
```diff
@@ -12,8 +12,8 @@ jobs:
     strategy:
       matrix:
         version:
-          - '1.3'
           - '1.6'
+          - '1.7'
```
Comment (devmotion) [inline .github/workflows/DynamicHMC.yml:16]:
```suggestion
          - '1'
```

File: .github/workflows/Numerical.yml [hunk 6]
```diff
@@ -12,8 +12,8 @@ jobs:
     strategy:
       matrix:
         version:
-          - '1.3'
           - '1.6'
+          - '1.7'
```
Comment (devmotion) [inline .github/workflows/Numerical.yml:16]:
```suggestion
          - '1'
```

File: .github/workflows/StanCI.yml [hunk 7]
```diff
@@ -12,8 +12,8 @@ jobs:
     strategy:
       matrix:
         version:
-          - '1.3'
           - '1.6'
+          - '1.7'
```
Comment (devmotion) [inline .github/workflows/StanCI.yml:16]:
```suggestion
          - '1'
```

File: .github/workflows/TuringCI.yml [hunk 8]
```diff
@@ -13,8 +13,8 @@ jobs:
     strategy:
       matrix:
         version:
-          - '1.3'
           - '1.6'
+          - '1.7'
```
Comment (devmotion) [inline .github/workflows/TuringCI.yml:17]:
```suggestion
          - '1'
```

**Pull request merged**